edx_user_state_client/tests.py

```diff
@@ -0,0 +1,349 @@
+
+    def test_set_two_blocks(self):
+        self.set(0, 0, {'a': 'b'})
+        self.set(0, 1, {'b': 'c'})
+        self.assertEquals(self.get(0, 0), {'a': 'b'})
+        self.assertEquals(self.get(0, 1), {'b': 'c'})
+
+    def test_set_many(self):
+        self.set_many(0, {0: {'a': 'b'}, 1: {'b': 'c'}})
+        self.assertEquals(self.get(0, 0), {'a': 'b'})
+        self.assertEquals(self.get(0, 1), {'b': 'c'})
+
+    def test_get_many(self):
+        self.set_many(0, {0: {'a': 'b'}, 1: {'b': 'c'}})
+        self.assertItemsEqual(
+            self.get_many(0, [0, 1]),
+            [
+                (self._block(0), {'a': 'b'}),
+                (self._block(1), {'b': 'c'})
+            ]
+        )
+
+    def test_delete(self):
+        with self.assertRaises(self.client.DoesNotExist):
+            self.get(0, 0)
+
+        self.set(0, 0, {'a': 'b'})
+        self.assertEqual(self.get(0, 0), {'a': 'b'})
+
+        self.delete(0, 0)
+        with self.assertRaises(self.client.DoesNotExist):
+            self.get(0, 0)
+
+    def test_delete_partial(self):
+        with self.assertRaises(self.client.DoesNotExist):
+            self.get(0, 0)
+
+        self.set(0, 0, {'a': 'b', 'b': 'c'})
+        self.assertEqual(self.get(0, 0), {'a': 'b', 'b': 'c'})
+
+        self.delete(0, 0, ['a'])
+        self.assertEqual(self.get(0, 0), {'b': 'c'})
+
+    def test_delete_last_field(self):
+        with self.assertRaises(self.client.DoesNotExist):
+            self.get(0, 0)
+
+        self.set(0, 0, {'a': 'b'})
+        self.assertEqual(self.get(0, 0), {'a': 'b'})
+
+        self.delete(0, 0, ['a'])
+        with self.assertRaises(self.client.DoesNotExist):
+            self.get(0, 0)
+
+    def test_delete_many(self):
+        self.assertItemsEqual(self.get_many(0, [0, 1]), [])
+
+        self.set_many(0, {
+            0: {'a': 'b'},
+            1: {'b': 'c'},
+        })
+
+        self.delete_many(0, [0, 1])
+        self.assertItemsEqual(self.get_many(0, [0, 1]), [])
+
+    def test_delete_many_partial(self):
+        self.assertItemsEqual(self.get_many(0, [0, 1]), [])
+
+        self.set_many(0, {
+            0: {'a': 'b'},
+            1: {'b': 'c'},
+        })
+
+        self.delete_many(0, [0, 1], ['a'])
+        self.assertItemsEqual(self.get_many(0, [0, 1]), [(self._block(1), {'b': 'c'})])
+
+    def test_delete_many_last_field(self):
+        self.assertItemsEqual(self.get_many(0, [0, 1]), [])
+
+        self.set_many(0, {
+            0: {'a': 'b'},
+            1: {'b': 'c'},
+        })
+
+        self.delete_many(0, [0, 1], ['a', 'b'])
+        self.assertItemsEqual(self.get_many(0, [0, 1]), [])
+
+
+class DictUserStateClient(XBlockUserStateClient):
+    """
+    The simplest possible in-memory implementation of DictUserStateClient,
+    for testing the tests.
+    """
+    def __init__(self):
+        self._data = {}
+
+    def get_many(self, username, block_keys, scope=Scope.user_state, fields=None):
```

Comment:
get raises DoesNotExist though get_many in this implementation does not. Obviously if you attempt to get many things some may not exist and you would get a partial response. The implementation feels uneven.